Cargo.toml

```diff
@@ -1,19 +1,28 @@
 [package]
 name = "cstr8"
-version = "0.1.2"
+version = "0.1.3"
```

Comment:
With no_std support added, I think it's viable to cut a 1.0 release. Thoughts?

I'd release a semver-hack 0.1.x re-export of ^1 ofc. It'd be nice to have this library at 1.0 so it's easier to justify its usage. Especially with `c""` literals expected to stabilize in 1.77.

Reply:
My initial look says "yes, 1.0 is good". I'll do a more thorough review soon. I have an idea for improving this macro further to accept `&str`, `&[u8]`, `&CStr`, and `[u8; N]` input, though it should be a non-breaking change.

Reply:
Somewhat interestingly, I was working with / writing about macro specialization earlier today, and built out the (annoying) machinery required to do autoderef specialization in a `const` context (i.e. where you can't directly call trait methods). I don't think I'll need it anytime soon (so probably won't implement it myself) but would be happy to review/merge an implementation. [Here's the playground where I got `const` autoderef specialization working.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1e59119988ec42a6c76edd5b3a8688cf)

Reply:
Nice playground, always glad to see more type inspection and `const` sorcery. #3 is much simpler than that; it's just many `const fn` with the same name in distinct `impl`s. I wasn't planning on exposing a trait bound for valid input types to the macro or for it to be extensible, so I don't think it needs to be more complex.

Upon further review, my primary concern with `CStr8` stabilization is that it exposes that it is backed by a `str`, via its `Deref` and `as_` methods, instead of being backed by a `CStr`. In contrast, `CString8` is backed by a `CString` (which is currently backed by a `Box<[u8]>`). 

That's a fine library choice to make, but it should be seriously reviewed. Is it more valuable to have these cheap impls, or to have ABI compatibility with `CStr`?

Reply:
Imo `CStr8` is a bit more "`str` but nul terminated" than it is "`CStr` but UTF-8". It also isolates `CStr8` from if `&CStr` ever manages to become a thin pointer. Native Rust code deals significantly more in `str` than `CStr` (being able to subslice is a huge benefit missed with nul-terminated strings) so it makes sense to privilege the Rust native choice instead of the FFI one.

The heap layout of both is identical, so it's fine that both `as_str()` and `as_c_str()` exist and both can exist whether `CStr8` wraps `str` or `CStr`. That `CStr8` uses `str` but `CString8` uses `CString` is more an accident of convenience than deliberate choice.

Reply:
I see the decision, though I'll admit that it is a little disappointing from an optimization perspective. If you're writing code that receives a raw pointer from C, you _must_ perform `strlen` in order to `unsafe`ly convert to `CStr8`. This seriously hampers what I can do safely _and_ efficiently with UTF-8 c-strings long-term and sort of defeats my use case for the crate.

Reply:
My "mediocre combination of both worlds" solution is to introduce a new intermediate type `NulStr`. `CStr8` would directly wrap `CStr` and provide `to_str` and `to_nul_str` methods. `NulStr` wraps `str` and derefs to it for cheap string operations, so a rename of what `CStr8` is today. That way,  you can preserve the "known to end in nul" state while also caching the length, allowing users to only do `strlen` in the exact places where it is needed.

But I can understand not wanting _another_ string type 😄 

> For Rust usage, the UTF-8 guarantee isn't particularly useful without the ability to get and use `&str`

You could say the same about `&CStr` and accessing the bytes via `&[u8]`. If this library were consistent with `CStr`, it would provide a `to_str` that performs the `strlen`, and _not_ eagerly deref to `str`. Any operations that don't require knowing the length up-front could be reimplemented to not need to convert to `&str` first.

> I'm hoping to do the erasable refactor soonish so maybe that works for your use case

It's a technically workable solution for my use case, though is less ergonomic than ideal. What's missing other than trait impls is `unsafe fn thin_from_ptr<'a>(p: *const i8) -> Thin<&'a CStr8>` - a way to construct a `Thin<&CStr8>` from a raw pointer without going through `CStr8::from_ptr`/`strlen` first. That'd be useful to have for `ffi::CStr` too!

If it ends up being a problem, the plan is to give [`utf8-cstr`](https://docs.rs/utf8-cstr/latest/utf8_cstr/struct.Utf8CStr.html) some TLC. I'll probably do so anyways because I think a "CStr that is UTF-8" type should exist as a dedicated type. This crate, despite its name, is distinctly _not_ that but rather a "str that is nul-terminated".

>  have some doubts that &std::ffi::CStr will actually become thin, despite the intent to leave that option open

Agree to disagree. `extern type` and its impacts like thin-`&CStr` are useful for so many cases that I believe its landing is inevitable and long-coming (like GATs). My hope is that post-monomorphization errors are seen as the more acceptable cost than additional rarely-used traits, panicking in `size_of_val`, or refusing to stabilize. It'd be a perfect-is-the-enemy-of-good decision like allowing both `panic`  in `const` and associated `const` in traits despite the risk to compile guarantees for generic code. And with the improvements to post-monomorphization errors for inline-const, this kind of error is becoming more and more acceptable.

> actually API-breaking change to make CStr not MetaSized

How is that an API-breaking change exactly? Pointer metadata isn't stable yet.